pkg/docker/config/config.go

```diff
@@ -142,9 +142,14 @@ func RemoveAuthentication(sys *types.SystemContext, registry string) error {
 	})
 }
 
-// RemoveAllAuthentication deletes all the credentials stored in auth.json
+// RemoveAllAuthentication deletes all the credentials stored in auth.json and kernel keyring
 func RemoveAllAuthentication(sys *types.SystemContext) error {
 	return modifyJSON(sys, func(auths *dockerConfigFile) (bool, error) {
+		// first try to remove credentials from kernel keyring
+		err := removeAllAuthFromKernelKeyring()
+		if err != nil {
+			logrus.Debugf("error removing credentials from kernel keyring")
+		}
```

Comment:
Should this _only_ remove data from the kernel keyring, or the data in the JSON as well?

The code as is removes _both_ `CredHelpers` and `AuthConfigs`, even if they both contain entries for the same registry; so it seems consistent for the keyring to be cleared as well as the two other locations.

OTOH, `RemoveAuthentication` does remove the data from exactly one location, not from all.

I don’t know. What do others think?

Reply:
podman logout -a will use RemoveAllAuthentication, so I think this should just clear all credentials from keyring, CredHelpers and AuthConfigs.

Reply:
Oops, I misread this for some reason; I thought the code is exiting on success. Sorry.

Nevertheless, let’s see what others think.

Reply:
Unless there's a clear mechanism to control which authentication mechanism to use, removing all sounds good to me.

Reply:
I think this related the criteria of the card [Podman: Complete kernel keyring auth](https://jira.coreos.com/browse/RUN-487?workflowName=Software+Simplified+Workflow+for+Project+TEMPLATE&stepId=6).
If `podman logout -a` need to control which storage to remove, c/image need to support the control.

Reply:
I want this PR only to complete the keyctl to remove all keyring. Just add function `removeAllAuthFromKernelKeyring()` without calling it.

There should be another PR to support Podman using keyring. `removeAllAuthFromKernelKeyring()` should be used in that PR. But I can't make the decision that how c/image support Podman using keyring. Add `AuthEnableKeyring` field to `SystemContext` ?

@rhatdan @mtrmac @vrothberg can you decide that?

Reply:
@QiWang19, once we have consensus, can you open a PR over at libpod vendoring this branch (i.e., `rm_all_keyring`)? I want to make sure that libpod's CI is green with the new changes before merging them into c/image.

> Add AuthEnableKeyring field to SystemContext ?

I support that as users are able to opt-in. If `AuthEnableKeyring == true` and the keyring doesn't work (e.g., privileges) we will still fall back to using the authfile, right?

Reply:
>If AuthEnableKeyring == true and the keyring doesn't work (e.g., privileges) we will still fall back to using the authfile
from this comment, no fallback. https://jira.coreos.com/browse/RUN-487?focusedCommentId=118874&page=com.atlassian.jira.plugin.system.issuetabpanels:comment-tabpanel#comment-118874
Podman will not use XDG_RUNTIME_DIR as default, no file to fall back

Reply:
Thanks for clarifying @QiWang19. @mtrmac WDYT?

Reply:
Ah, we do currently have the `enableKeyring` bool. I suppose, then, that this PR should make the code conditional on that bool as well, to be consistent with everything else.

We can then discuss the public API for opting in/out in a separate PR.

Does that work?

Reply:
Works for me :+1: